usr/src/ka9q-1.4/xmessage.src/xmessage.man
.TH XMESSAGE 1 "20 August 1991" "X Version 11"
.SH NAME
xmessage \- display a message or query in a window (X-based /bin/echo)
.SH SYNOPSIS
.B xmessage
[
.B \-buttons
.I label1\fR[\fP:value1\fR]\fP\fB,\fPlabel2\fR[\fP:value2\fR]\fP\fB,\fP\|.\|.\|.
] [
.I options
]
.B \-file
.I filename
.br
.B xmessage
[
.B \-buttons
.I label1\fR[\fP:value1\fR]\fP\fB,\fPlabel2\fR[\fP:value2\fR]\fP\fB,\fP\|.\|.\|.
] [
.I options
]
.I message
\&.\|.\|.
.SH DESCRIPTION
.I Xmessage
displays a window containing a message from the command line, a file,
or standard input.
Along the lower edge of the message is row
of buttons; clicking the left mouse button
on any of these buttons will cause \fIxmessage\fP to exit.
Which button was pressed is returned in the exit status and,
optionally, by writing the label of the button to standard output.
.PP
.I Xmessage
is typically used by shell scripts to display information to the user
or to ask the user to make a choice.
.SH OPTIONS
These are the command line options that \fIxmessage\fP understands.
.TP 8
.B \-buttons \fIbutton,button,.\|.\|.\fP
This option will cause \fIxmessage\fP to create one button for each
comma-separated \fIbutton\fP argument.
The corresponding resource is \fBbuttons\fP.
Each \fIbutton\fP consists of a label optionally followed by a colon
and an exit value.
The label is the name of the Command button widget created and will be the 
default text displayed to the user.  Since this is the name of the widget
it may be used to change any of the resources associated with that button.
The exit value will be returned by \fIxmessage\fP
if that button is selected.  The default exit value is 100 plus the
button number.  Buttons are numbered from the left starting with one.
The default string if no \fB\-buttons\fP option is given is \fBokay:0\fP.
.TP 8
.B \-default \fIlabel\fP
Defines the button with a matching \fIlabel\fP to be the default.
If not specified there is no default.
The corresponding resource is \fBdefaultButton\fP.
Pressing Return anywhere in the \fIxmessage\fP window will activate
the default button.
The default button has a wider border than the others.
.TP 8
.B \-file \fIfilename\fP
File to display.
The corresponding resource is \fBfile\fP.
A \fIfilename\fP of `\fB\-\fP' reads from standard input.
If this option is not supplied, \fIxmessage\fP will
display all non-option arguments in the style of \fIecho\fP.
Either \fB\-file\fP or a message on the command line
should be provided, but not both.
.TP 8
.B \-print
This will cause the program to print the label of the button pressed to
standard output.  Equivalent to setting the \fBprintValue\fP resource
to TRUE.
This is one way to get feedback as to which button was pressed.
.SH "WIDGET HIERARCHY"
Knowing the name and position in the hierarchy of each widget is
useful when specifying resources for them.  In the chart below, the
class and name of each widget is given.
.LP
.nf
Xmessage (xmessage)
	Form form
		Label message
		Command (label1)
		Command (label2)
		.
		.
		.
.fi
.SH RESOURCES
\fIXmessage\fP has a few top-level application resources that allow
customizations that are specific to \fIxmessage\fP. 
.TP 8
.B file
A String resource specifying the file to display.
.TP 8
.B buttons
A String resource specifying the buttons to display.
See the \fB\-buttons\fP command-line option.
.TP 8
.B defaultButton
A String resource specifying a default button by label.
.TP 8
.B printValue
A Boolean resource that determines whether or not the label of the 
button pressed to exit the program is printed.  The default is FALSE.
.SH ACTIONS
.TP 8
.B exit(\fIvalue\fP)
exit immediately with an exit status of \fIvalue\fP (default 0).  This
action can be used with translations to provide alternate ways of
exiting \fIxmessage\fP.
.TP 8
.B default-exit()
exit immediately with the exit status specified by the default button.
If there is no default button, this action has no effect.
.SH "EXIT STATUS"
If it detects an error, \fIxmessage\fP returns 1, so this value should
not be used with a button.
.SH "SEE ALSO"
.IR X (1),
.IR echo (1),
.IR cat (1)
.SH COPYRIGHT
Copyright 1988,1991 Massachusetts Institute of Technology.
.br
See \fIX\fP(1) for a full statement of rights and permissions.
.SH AUTHORS
Chris Peterson, MIT Project Athena 
.br
Stephen Gildea, MIT X Consortium


usr/src/ka9q-1.4/xmessage.src/xmessage.c
/*
 * xmessage - utility for querying users
 *
 * Copyright 1988,1991 Massachusetts Institute of Technology
 * Time-stamp: <91/08/27 18:33:26 gildea>
 *
 * Permission to use, copy, modify, and distribute this software and its
 * documentation for any purpose and without fee is hereby granted, provided
 * that the above copyright notice appear in all copies and that both that
 * copyright notice and this permission notice appear in supporting
 * documentation, and that the name of M.I.T. not be used in advertising or
 * publicity pertaining to distribution of the software without specific,
 * written prior permission.  M.I.T. makes no representations about the
 * suitability of this software for any purpose.  It is provided "as is"
 * without express or implied warranty.
 *
 * Originally written by Chris D. Peterson, MIT Project Athena, 1988.
 * Updated by Jim Fulton, MIT X Consortium.
 * Current version by Stephen Gildea <gildea@expo.lcs.mit.edu>, August 1991.
 */

#include <X11/Intrinsic.h>
#include <X11/StringDefs.h>
#include <stdio.h>

extern char *malloc();

extern char *read_file();
extern Widget make_queryform();

/*
 * data used by xmessage
 */

char *ProgramName;

static Atom wm_delete_window;

static struct _QueryResources {
    char *file;
    char *button_list;
    char *default_button;
    Boolean print_value;
    char *message;
} qres;				/* initialized by resources below */

#define offset(field) XtOffsetOf(struct _QueryResources, field)
static XtResource resources[] = {
    { "file", "File", XtRString, sizeof (char *), 
      offset(file), XtRString, (XtPointer) NULL },
    { "buttons", "Buttons", XtRString, sizeof (char *),
      offset(button_list), XtRString, (XtPointer) "okay:0" },
    { "defaultButton", "DefaultButton", XtRString, sizeof (char *),
      offset(default_button), XtRString, (XtPointer) NULL },
    { "printValue", "PrintValue", XtRBoolean, sizeof (Boolean),
      offset(print_value), XtRString, "false" },
};
#undef offset

static XrmOptionDescRec optionList[] =  {
    { "-file",    ".file", XrmoptionSepArg, (XPointer) NULL },
    { "-buttons", ".buttons", XrmoptionSepArg, (XPointer) NULL },
    { "-default", ".defaultButton", XrmoptionSepArg, (XPointer) NULL },
    { "-print",   ".printValue", XrmoptionNoArg, (XPointer) "on" },
};

static String fallback_resources[] = {
    "*baseTranslations: #override :<Key>Return: default-exit()",
    NULL};


/*
 * usage
 */

static void usage ()
{
    static char *options[] = {
"    -file filename              file to read message from, - means stdin",
"    -buttons string             comma-separated list of label:exitcode",
"    -default button             button to activate if Return is pressed",
"    -print                      print the button label when selected",
"",
NULL};
    char **cpp;

    fprintf (stderr, "usage: %s [-options] [message ...]\n\n",
	     ProgramName);
    fprintf (stderr, "where options include:\n");
    for (cpp = options; *cpp; cpp++) {
	fprintf (stderr, "%s\n", *cpp);
    }
    exit (1);
    /* NOTREACHED */
}

/*
 * Action to implement ICCCM delete_window and other translations.
 * Takes one argument, the exit status.
 */
/* ARGSUSED */
static void
exit_action(w, event, params, num_params)
    Widget w;			/* unused */
    XEvent *event;
    String *params;
    Cardinal *num_params;
{
    int exit_status = 0;

    if(event->type == ClientMessage
       && event->xclient.data.l[0] != wm_delete_window)
	return;
    
    if (*num_params == 1)
	exit_status = atoi(params[0]);
    exit(exit_status);
}

int default_exitstatus = -1;		/* value of button named by -default */

/* ARGSUSED */
static void
default_exit_action(w, event, params, num_params)
    Widget w;
    XEvent *event;
    String *params;
    Cardinal *num_params;
{
    if (default_exitstatus >= 0)
	exit(default_exitstatus);
}

XtActionsRec actions_list[] = {
    "exit", exit_action,
    "default-exit", default_exit_action,
};

static String top_trans =
    "<ClientMessage>WM_PROTOCOLS: exit(1)\n";

/*
 * xmessage main program - make sure that there is a message,
 * then create the query box and go.  Callbacks take care of exiting.
 */
main (argc, argv)
    int argc;
    char **argv;
{
    Widget top, queryform;
    XtAppContext app_con;

    ProgramName = argv[0];

    top = XtAppInitialize (&app_con, "Xmessage",
			   optionList, XtNumber(optionList), &argc, argv,
			   fallback_resources, NULL, 0);

    XtGetApplicationResources (top, (XtPointer) &qres, resources,
			       XtNumber(resources), NULL, 0);

    if (argc > 1 && !strcmp(argv[1], "-help")) {
	usage();
    }
    if (argc == 1 && qres.file != NULL) {
	qres.message = read_file (qres.file);
	if (qres.message == NULL) {
	    fprintf (stderr, "%s: problems reading message file\n",
		     ProgramName);
	    exit (1);
	}
    } else if (argc > 1 && qres.file == NULL) {
	int i, len;
	char *cp;

	len = argc - 1;		/* spaces between words and final NULL */
	for (i=1; i<argc; i++)
	    len += strlen(argv[i]);
	qres.message = malloc(len);
	if (!qres.message)
	    perror("string malloc");
	cp = qres.message;
	for (i=1; i<argc; i++) {
	    strcpy(cp, argv[i]);
	    cp += strlen(argv[i]);
	    if (i != argc-1)
		*cp++ = ' ';
	    else
		*cp = '\0';
	}
    } else {
	usage ();
    }

    if (qres.message == NULL) {
	fprintf (stderr, "%s: no message\n", ProgramName);
	exit (1);
    }

    XtAppAddActions(app_con, actions_list, XtNumber(actions_list));
    XtOverrideTranslations(top, XtParseTranslationTable(top_trans));

    /*
     * create the query form; this is where most of the real work is done
     */
    queryform = make_queryform (top, qres.message, qres.button_list,
				qres.print_value, qres.default_button);
    if (!queryform) {
	fprintf (stderr,
		 "%s: unable to create query form with buttons: %s\n",
		 ProgramName, qres.button_list);
	exit (1);
    }

    XtSetMappedWhenManaged(top, FALSE);
    XtRealizeWidget(top);

    /* do WM_DELETE_WINDOW before map */
    wm_delete_window = XInternAtom(XtDisplay(top), "WM_DELETE_WINDOW", False);
    XSetWMProtocols(XtDisplay(top), XtWindow(top), &wm_delete_window, 1);

    XtMapWidget(top);
    XtAppMainLoop(app_con);

    exit (0);
}
